lib/node/node.js

```diff
@@ -449,7 +457,11 @@ module.exports.create = function createNode(type, expectedParams, options) {
         // `beforeCreate` allows to change/modify the node before it's actually created/returned
         // if it throws an error it will fail to create the node.
         if (options.beforeCreate) {
-            options.beforeCreate.bind(this)();
+            this.beforeCreate = options.beforeCreate.bind(this);
```

Comment:
Why don't we make everything async?

I suppose is because we don't want to change all the `beforeCreate` methods but to have two methods is a bit confusing and dangerous also as @dgaubert always say we don't want to [release the zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony)

What do you think? We can leave it this way and create a ticket to change it in the future?

Reply:
Hey!

Let me clarify some misconceptions about __releasing Zalgo__.  

### Context
 
We have a class hierarchy between `node` and child implementations such as `GeoreferenceCity` or `source`. 

This PR introduces in `node` the method `beforeCreate` and is intended to be an abstract method with a default implementation (we had it already... implicitly 😄). That method is defined as *sync method* and then every child can implement its own version of `beforeCreate`.

Pretty much the same with `beforeCreateAsync`, here we add a `beforeCreateAsync` abstract method and every child-node can overwrite it. 

The problem here is that the validator calls to `beforeCreate` synchronously during validation. So we don't want to know the details of that method. Then `beforeCreate` must be defined as synchronous method.

On the other hand, `beforeCreateAsync` is called by `analysis-factory` and is intended, as @rochoa said, to fetch some metadata from database. Then `beforeCreateAsync` must be defined as asynchronous method.

If we implement `beforeCreate` expecting a callback and validator never provides it then analysis creation will never end. Viceversa with `beforeCreateAsync`, if we don't provide a callback while calling `beforeCreateAsync` a `undefined is not a function` will be thrown.

### Releasing zalgo

So far, we don't have to worry about `releasing zalgo`. The problem will come if we implements `beforeCreateAsync`, for example, calling the callback asynchronously when everything goes right and calling it synchronously when something goes wrong, usually it appears while checking some condition before calling to the inner async function, i.e:

```
function foo(bar, callback) {
    if (!bar) return callback(new Error('Ooops')); // called synchronously
    baz(bar, function (err, qux) {
         if (err) {
             return callback(err); // called asynchronously
         }
         callback(null, qux); // called asynchronously
    });
}
```

You may notice that we call callbacks synchronously a lot of times in our projects. Actually, it's not so bad thing, only under certain conditions we can block the event loop, such as intensive recursion and so on. Also most of those functions are not public APIs. However, it's a good thing to be aware of it.

Hope this helps.

Reply:
Understood, thank you for this amazing explanation :)